test/cpp/util/cli_credentials.cc

```diff
@@ -29,47 +31,107 @@ DEFINE_string(
     ssl_target, "",
     "If not empty, treat the server host name as this for ssl/tls certificate "
     "validation.");
+DEFINE_string(channel_creds_type, "",
+              "The channel creds type: insecure, ssl, or alts.");
 
 namespace grpc {
 namespace testing {
 
-std::shared_ptr<grpc::ChannelCredentials> CliCredentials::GetCredentials()
+grpc::string CliCredentials::GetDefaultChannelCredsType() const {
+  // Compatibility logic for --enable_ssl.
+  if (FLAGS_enable_ssl) {
+    fprintf(stderr,
+            "warning: --enable_ssl is deprecated. Use "
+            "--channel_creds_type=ssl.\n");
+    return "ssl";
+  }
+  // Implicit channel for GoogleDefaultCredentials is SSL.
+  if (FLAGS_access_token.empty() && FLAGS_use_auth) {
+    return "ssl";
+  }
+  return "insecure";
+}
+
+std::shared_ptr<grpc::ChannelCredentials>
+CliCredentials::GetChannelCredentials() const {
+  if (FLAGS_channel_creds_type.compare("insecure") == 0) {
+    return grpc::InsecureChannelCredentials();
+  } else if (FLAGS_channel_creds_type.compare("ssl") == 0) {
+    return grpc::SslCredentials(grpc::SslCredentialsOptions());
+  } else if (FLAGS_channel_creds_type.compare("alts") == 0) {
+    return grpc::experimental::AltsCredentials(
+        grpc::experimental::AltsCredentialsOptions());
+  }
+  fprintf(stderr,
+          "--channel_creds_type=%s invalid; must be insecure, ssl or alts.\n",
+          FLAGS_channel_creds_type.c_str());
+  return std::shared_ptr<grpc::ChannelCredentials>();
+}
+
+std::shared_ptr<grpc::CallCredentials> CliCredentials::GetCallCredentials()
     const {
   if (!FLAGS_access_token.empty()) {
     if (FLAGS_use_auth) {
       fprintf(stderr,
               "warning: use_auth is ignored when access_token is provided.");
     }
-
-    return grpc::CompositeChannelCredentials(
-        grpc::SslCredentials(grpc::SslCredentialsOptions()),
-        grpc::AccessTokenCredentials(FLAGS_access_token));
+    return grpc::AccessTokenCredentials(FLAGS_access_token);
   }
+  // TODO(@capstan): Support GoogleDefaultCredentials on other channel types.
```

Comment:
I do not understand this todo.

Reply:
Given that GDC are essentially SSL + local auth credentials as pulled from the local environment (ENV variable + local file with JWT, AppEngine Metadata API, GCE Metadata API), there's no reason that said credentials could be used on a different channel type. That said, GDC's creds accessor is only a ChannelCreds, and thus cannot be used with ALTS or LOAS, as you can only pick one ChannelCreds and composite CallCreds on top of it.

If the answer is that GDC is supposed to *be* the equivalent of transport credentials for transports that don't have them, then perhaps this TODO is not meant to be achievable. If so, I should update this and downstream PRs to move `google_default_credentials` to be a `channel_creds_type`.

Reply:
My understanding is GDC is a default transport credentials which should use the proper channel and call creds type depending on where it is running. Potentially it means it can be alts when running in gce and such (even though it is just ssl + call creds as you said today).

Reply:
Ok, brought `google_default_credentials` in as a channel type. PTAL.